File: 0002.add_two_numbers/solution.go
package solution

/**
 * Definition for singly-linked list.
 */

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	head := &ListNode{}
	prev := head
	carry, curVal := 0, 0

	for l1 != nil && l2 != nil {
		val := l1.Val + l2.Val + carry
		curVal, carry = val%10, val/10
		cur := &ListNode{Val: curVal}
		prev.Next, prev, l1, l2 = cur, cur, l1.Next, l2.Next
	}

	for l1 != nil {
		val := l1.Val + carry
		curVal, carry = val%10, val/10
		cur := &ListNode{Val: curVal}
		prev.Next, prev, l1 = cur, cur, l1.Next
	}

	for l2 != nil {
		val := l2.Val + carry
		curVal, carry = val%10, val/10
		cur := &ListNode{Val: curVal}
		prev.Next, prev, l2 = cur, cur, l2.Next
	}

	if carry != 0 {
		prev.Next = &ListNode{Val: carry}
	}

	return head.Next
}

File: 0002.add_two_numbers/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode head = new ListNode(0);
        ListNode prev = head;
        int carry = 0, curVal = 0;

        while (l1 != null && l2 != null) {
            int val = l1.val + l2.val + carry;
            carry = val / 10;
            curVal = val % 10;

            ListNode cur = new ListNode(curVal);

            prev.next = cur;
            prev = cur;
            l1 = l1.next;
            l2 = l2.next;
        }

        while (l1 != null) {
            int val = l1.val + carry;
            carry = val / 10;
            curVal = val % 10;

            ListNode cur = new ListNode(curVal);

            prev.next = cur;
            prev = cur;
            l1 = l1.next;
        }

        while (l2 != null) {
            int val = l2.val + carry;
            carry = val / 10;
            curVal = val % 10;

            ListNode cur = new ListNode(curVal);

            prev.next = cur;
            prev = cur;
            l2 = l2.next;
        }

        if (carry != 0) {
            prev.next = new ListNode(carry);
        }

        return head.next;
    }
}
